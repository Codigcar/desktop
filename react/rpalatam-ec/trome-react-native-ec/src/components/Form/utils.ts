import { useCallback, useState } from 'react'

import { FormErrorsProps, FormErrorsReturn, ValidationErros } from './types'

export const useFormErrors = ({
  ref: formRef,
  schema,
}: FormErrorsProps): FormErrorsReturn => {
  const [validationErrors, setValidationErros] = useState<ValidationErros>({})

  const handleChangeField = useCallback(
    async (name, value) => {
      const error = formRef.current?.getFieldError(name)
      const hasError = error && error.length > 0
      try {
        await schema.validate({ [name]: value })
        if (hasError) {
          formRef.current?.setFieldError(name, '')
          setValidationErros((prev) => ({ ...prev, [name]: '' }))
        }
      } catch (err) {
        if (err.inner[name] && (!error || error !== err.inner[name])) {
          formRef.current?.setFieldError(name, err.inner[name])
          setValidationErros((prev) => ({ ...prev, [name]: err.inner[name] }))
        } else if (!err.inner[name] && hasError) {
          formRef.current?.setFieldError(name, '')
          setValidationErros((prev) => ({ ...prev, [name]: '' }))
        }
      }
    },
    [formRef, schema],
  )

  const setErrors = useCallback(
    (errors) => {
      formRef.current?.setErrors(errors)
      setValidationErros(errors)
    },
    [formRef],
  )

  return {
    errors: validationErrors,
    setFieldError: handleChangeField,
    setErrors,
  }
}
