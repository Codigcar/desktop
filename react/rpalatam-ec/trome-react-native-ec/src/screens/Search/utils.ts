export const computeEstimatedDate = (date: number): string => {
  const currentDate = new Date()
  const approx = currentDate.getDate() + date
  currentDate.setDate(approx)

  const month = `0${currentDate.getUTCMonth() + 1}`.slice(-2)
  const day = `0${currentDate.getUTCDate()}`.slice(-2)
  const year = currentDate.getUTCFullYear()

  return `${year}-${month}-${day}`
}

export const parseDate = (date?: Date): string | null => {
  if (!date) return null
  const month = `0${date.getUTCMonth() + 1}`.slice(-2)
  const day = `0${date.getUTCDate()}`.slice(-2)
  const year = date.getUTCFullYear()

  return `${year}-${month}-${day}`
}

export const timeFrames = [
  {
    value: '',
    label: 'Cualquier fecha',
  },
  {
    value: '-1',
    label: 'Últimas 24 horas',
  },
  {
    value: '-7',
    label: 'Última semana',
  },
  {
    value: '-30',
    label: 'Último mes',
  },
  {
    value: 'custom',
    label: 'Personalizado',
  },
]
