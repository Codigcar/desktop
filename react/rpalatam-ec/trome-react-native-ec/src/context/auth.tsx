import analytics from '@react-native-firebase/analytics'
import crashlytics from '@react-native-firebase/crashlytics'
import { EventEmitter } from 'events'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react'
import { Linking } from 'react-native'

import { User } from '../entities/User'
import auth from '../services/auth'
import subscription from '../services/subscription'
import { googleSignOut, init, tokenExchange } from '../utils/auth'
import { ENABLE_AUTHENTICATION } from '../utils/flags'
import { sendPostMessageToWebview } from '../utils/refs'

type AuthListener = (isAuthenticated: boolean) => void

interface Context {
  isAuthenticated: boolean | undefined
  isSubscribed: boolean | undefined
  signin: (provider: string) => void
  signout: () => Promise<void>
  updateUserProfile: (userProfile: User) => Promise<void>
  onAuthenticationChange: (listener: AuthListener) => () => void
  user: User
  verifySubscription: () => Promise<void>
}

const AuthContext = createContext<Context>({} as Context)
const authEmitter = new EventEmitter()
const AUTH_EVENT_NAME = 'auth.event'

export const AuthProvider: React.FC = ({ children }) => {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>()
  const [user, setUser] = useState<User>({})
  const [subscriptor, setSubscriptor] = useState<boolean>()

  const onAuthenticationChange = useCallback((listener: AuthListener) => {
    authEmitter.on(AUTH_EVENT_NAME, listener)
    return function () {
      authEmitter.off(AUTH_EVENT_NAME, listener)
    }
  }, [])

  useEffect(() => {
    init()
    const restoreSession = async () => {
      try {
        if (!ENABLE_AUTHENTICATION) return setIsAuthenticated(false)
        const token = await auth.getToken()
        if (!token) return setIsAuthenticated(false)
        setIsAuthenticated(true)
      } catch (error) {
        if (error instanceof Error) crashlytics().recordError(error)
        setIsAuthenticated(false)
      }
    }
    restoreSession()
  }, [])

  const signin = useCallback((_: string) => {
    setIsAuthenticated(true)
    authEmitter.emit(AUTH_EVENT_NAME, true)
  }, [])

  const signout = useCallback(async () => {
    await auth.logout()
    await googleSignOut()
    await analytics().logEvent('logout')
    setIsAuthenticated(false)
    sendPostMessageToWebview('Home', { type: 'auth.LOGOUT' })
    authEmitter.emit(AUTH_EVENT_NAME, false)
  }, [])

  const updateUserProfile = useCallback(async (data: User) => {
    sendPostMessageToWebview('Home', {
      type: 'auth.PROFILE',
      payload: { ...data },
    })
    setUser(data)
  }, [])

  useEffect(() => {
    if (typeof isAuthenticated === 'undefined') return
    if (isAuthenticated) {
      auth
        .getUserProfile()
        .then((profile) => {
          sendPostMessageToWebview('Home', {
            type: 'auth.PROFILE',
            payload: { ...profile },
          })
          setUser(profile)
        })
        .catch((error) => {
          signout()
          crashlytics().recordError(error)
        })
    } else {
      setUser({})
    }
  }, [isAuthenticated, signout])

  const verifySubscription = useCallback(async () => {
    if (!subscription) return
    try {
      const params: Record<string, string | undefined> = { id: user.id }
      const status = await subscription.checkAccess(params)
      setSubscriptor(status)
    } catch (error) {
      if (error instanceof Error) crashlytics().recordError(error)
    }
  }, [user.id])

  useEffect(() => {
    if (typeof isAuthenticated === 'undefined') return
    if (isAuthenticated && !user.id) return
    if (isAuthenticated && user.id) {
      verifySubscription()
    } else {
      setSubscriptor(false)
    }
  }, [isAuthenticated, user.id, verifySubscription])

  useEffect(() => {
    if (typeof isAuthenticated === 'undefined') return
    if (isAuthenticated && !user.id) return
    const id = user.id || null
    analytics().setUserId(id)
    crashlytics().setUserId(id || '')
  }, [isAuthenticated, user.id])

  useEffect(() => {
    if (typeof subscriptor === 'undefined') return
    const type = subscriptor ? 'premium' : 'free'
    analytics().setUserProperty('account_type', type)
    sendPostMessageToWebview('Home', {
      type: 'subscription.UPDATE_STATUS',
      payload: { status: subscriptor },
    })
  }, [subscriptor])

  useEffect(() => {
    const callback = async (params: { url: string }) => {
      try {
        const config = auth.getOAuthConfig?.()
        if (!config) return
        if (!params.url.startsWith(config.redirect_url)) return
        const token = await tokenExchange(config, params.url)
        await auth.setToken(token)
        signin('oauth')
      } catch (_) {}
    }
    const listener = Linking.addEventListener('url', callback)
    return () => {
      listener.remove()
    }
  }, [signin])

  return (
    <AuthContext.Provider
      value={{
        isAuthenticated,
        isSubscribed: subscriptor,
        signin,
        signout,
        updateUserProfile,
        onAuthenticationChange,
        user,
        verifySubscription,
      }}>
      {children}
    </AuthContext.Provider>
  )
}

export const useAuth = (): Context => useContext(AuthContext)
