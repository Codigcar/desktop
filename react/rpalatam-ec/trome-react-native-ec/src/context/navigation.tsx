import AsyncStorage from '@react-native-async-storage/async-storage'
import analytics from '@react-native-firebase/analytics'
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useRef,
  useState,
} from 'react'

import * as utils from '../utils/categories'
import { STORE_CATEGORIES } from '../utils/constants'
import { NavigationItem, getMainNavigation } from '../utils/navigation'
import { HomeWebviewRef } from '../utils/refs'
import type { NavCategory } from '../utils/categories'

type SetListOfCategory =
  | NavCategory[]
  | ((prevState: NavCategory[]) => NavCategory[])

type Context = {
  categories: NavCategory[]
  mainNavigation: NavigationItem[]
  sendCategoriesToWebview: () => void
  setListOfCategory: (categories: SetListOfCategory) => Promise<void>
}

const NavigationContext = createContext<Context>({} as Context)

export const NavigationProvider: React.FC = ({ children }) => {
  const [mainNavigation, setMainNavigation] = useState<NavigationItem[]>([])
  const [categories, setCategories] = useState<NavCategory[]>([])
  const refCategories = useRef<NavCategory[]>([])

  const sendCategoriesToWebview = useCallback(() => {
    HomeWebviewRef.current?.injectJavaScript(`(function() {
      if (window.NATIVE_CONNECTION && window.NATIVE_CONNECTION.categories) {
        window.NATIVE_CONNECTION.categories.setCategoriesConfig(${JSON.stringify(
          categories,
        )});
      }
    })();`)
  }, [categories])

  const setListOfCategory = useCallback(
    async (listOfCategory: SetListOfCategory) => {
      const { current } = refCategories
      const isFn = typeof listOfCategory === 'function'
      const list = isFn ? listOfCategory(current) : listOfCategory

      const changed = list.find((category) => {
        const item = current.find(({ key }) => key === category.key)
        if (!item) return false
        return !item.active !== !category.active
      })
      if (changed) {
        const name = changed.active ? 'add_to_home' : 'remove_from_home'
        await analytics().logEvent(name, {
          id: changed.key,
          name: changed.label,
          type: 'section',
        })
      }

      await AsyncStorage.setItem(STORE_CATEGORIES, JSON.stringify(list))
      setCategories(list)
    },
    [],
  )

  const initNavigation = useCallback(async () => {
    const [navigation, localCategories] = await Promise.all([
      getMainNavigation(),
      utils.loadLocalCategories(),
    ])
    setMainNavigation(navigation)

    const categoryList = navigation
      .filter((item) => item.type === 'category')
      .map((item, index) => ({
        active: index < 7,
        required: item.key === 'portada',
        ...item,
      }))

    if (refCategories.current.length === 0 && !localCategories) {
      setCategories(categoryList)
      return
    }

    const composedCategories = utils.composeCategoriesWithLocal(
      categoryList,
      localCategories || refCategories.current,
    )
    setCategories(composedCategories)
  }, [])

  useEffect(() => {
    initNavigation()
  }, [initNavigation])

  useEffect(() => {
    if (categories.length > 0) sendCategoriesToWebview()
    refCategories.current = categories
  }, [categories, sendCategoriesToWebview])

  return (
    <NavigationContext.Provider
      value={{
        categories,
        mainNavigation,
        sendCategoriesToWebview,
        setListOfCategory,
      }}>
      {children}
    </NavigationContext.Provider>
  )
}

export const useMainNavigation = (): Context => useContext(NavigationContext)
