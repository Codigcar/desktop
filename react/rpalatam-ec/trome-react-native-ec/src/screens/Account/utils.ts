import { App } from '../../utils/config'
import * as Schema from '../../utils/validation'

const name = Schema.string()
  .matches(/\d+/i, 'Este campo no puede contener números')
  .matches(/\D{51,}/i, 'Máximo 50 caracteres')

export const ProfileSchema = Schema.object({
  first_name: App.select({
    elcomercio: name.clone().min(2).required(),
    default: name.clone().required(),
  }),
  last_name: name.clone().min(2).required(),
  second_last_name: name.clone().min(2),
  email: Schema.string().required().email(),
  mobile_phone: App.select({
    elcomercio: Schema.string()
      .matches(/\D+/i, 'Se aceptan solo números')
      .min(6, 'Mínimo 6 dígitos')
      .max(12, 'Máximo 12 dígitos'),
    default: Schema.string()
      .min(9, 'Mínimo 9 dígitos')
      .max(13, 'Máximo 13 dígitos'),
  }),
  user_metadata: {
    documentNumber: App.select({
      elcomercio: Schema.string().required(),
      default: Schema.string(),
    })
      .matches(/\s+/i, 'No se permiten espacios en blanco')
      .when('user_metadata.documentType', (value, schema) => {
        if (value === undefined) return schema
        if (value === 'DNI') {
          return schema
            .required()
            .matches(/\D+/i, 'Se aceptan solo números')
            .min(8, 'Se aceptan 8 dígitos')
            .max(8, 'Se aceptan 8 dígitos')
        }
        return schema
          .required()
          .min(9, 'Mínimo 9 dígitos')
          .max(20, 'Máximo 20 dígitos')
      }),
    documentType: Schema.string(),
  },
})
const gendersBase = [
  { label: 'Hombre', value: 'MALE' },
  { label: 'Mujer', value: 'FEMALE' },
]
export const OPTIONS = {
  country: [{ label: 'Perú', value: '260000' }],
  civilStatus: [
    { label: 'Soltero (a)', value: 'SO' },
    { label: 'Casado (a)', value: 'CA' },
    { label: 'Divorciado (a)', value: 'DI' },
    { label: 'Viudo (a)', value: 'VI' },
  ],
  document: [
    { label: 'DNI', value: 'DNI' },
    { label: 'CEX', value: 'CEX' },
    { label: 'CDI', value: 'CDI' },
  ],
  gender: App.select({
    elcomercio: [
      ...gendersBase,
      { label: 'Prefiero no decirlo', value: 'UNKNOWN' },
      { label: 'Otro', value: 'OTHER' },
    ],
    default: gendersBase,
  }),
}
