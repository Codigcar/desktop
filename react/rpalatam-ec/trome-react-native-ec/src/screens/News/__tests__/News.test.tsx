import { act, cleanup, fireEvent, render } from '@testing-utils/library'
import MockAxios from 'jest-mock-axios'
import React from 'react'
import { Alert } from 'react-native'
import { Banner } from 'react-native-ad-manager'

import { useAuth } from '../../../context/auth'
import { FavoritesProvider } from '../../../context/favorites'
import { NavigationProvider } from '../../../context/navigation'
import { useNotification } from '../../../context/notification'
import { App } from '../../../utils/config'
import NewsScreen from '../News'

const Container = ({ children }: { children: JSX.Element }) => (
  <NavigationProvider>
    <FavoritesProvider>{children}</FavoritesProvider>
  </NavigationProvider>
)

const MOCK_RESPONSE = {
  data: {
    content_elements: [
      {
        _id: '1',
        canonical_url: 'url',
        headlines: { basic: 'title_one' },
        last_updated_date: 'date',
        subheadlines: { basic: 'description_one' },
        type: 'story',
      },
      {
        _id: '2',
        canonical_url: 'url',
        headlines: { basic: 'title_two' },
        last_updated_date: 'date',
        subheadlines: { basic: 'description_two' },
        type: 'story',
      },
    ],
    type: 'results',
  },
}

jest.mock('@react-navigation/native', () => ({
  useNavigation: jest.fn(() => ({
    navigate: jest.fn(),
    push: jest.fn(),
  })),
  useRoute: jest.fn(() => ({
    params: {
      section: {
        name: 'politics',
        path: 'politics',
      },
    },
  })),
}))

jest.mock('../../../context/auth')
const mockUseAuth = useAuth as jest.Mock

jest.mock('../../../context/notification')
const mockUseNotification = useNotification as jest.Mock

beforeEach(() => {
  mockUseAuth.mockImplementation(() => ({
    isSubscribed: false,
    user: {},
  }))
  mockUseNotification.mockImplementation(() => ({
    show: jest.fn(),
  }))
})

afterEach(() => {
  MockAxios.reset()
  cleanup()
})

describe('News screen page', () => {
  it('should render card and ad', async () => {
    const { getByTestId, getAllByTestId, getByText, toJSON } = render(
      <Container>
        <NewsScreen />
      </Container>,
    )
    expect(getAllByTestId('placeholder')).toBeDefined()
    await act(async () => undefined)
    await act(async () => MockAxios.mockResponse(MOCK_RESPONSE))

    const title = getByText('title_one')
    expect(title).toBeDefined()
    fireEvent.press(title)

    const screen = getByTestId('News-screen')
    // eslint-disable-next-line testing-library/await-async-query
    expect(screen.findByType(Banner)).toBeDefined()
    expect(toJSON()).toMatchSnapshot()
  })

  it('should not render ad when user is premium', async () => {
    mockUseAuth.mockImplementation(() => ({
      isSubscribed: true,
      user: { id: 'id' },
    }))

    const { getByTestId, getByText } = render(
      <Container>
        <NewsScreen />
      </Container>,
    )
    await act(async () => MockAxios.mockResponse(MOCK_RESPONSE))

    expect(getByText('title_one')).toBeDefined()

    const screen = getByTestId('News-screen')
    try {
      // eslint-disable-next-line testing-library/await-async-query
      screen.findByType(Banner)
      expect(true).toBe(false)
    } catch (error) {
      const message = 'No instances found with node type: "Banner"'
      expect((error as Error).message).toBe(message)
    }
  })

  it('should change icon when section is added or removed', async () => {
    App.key = 'gestion'
    const { getByTestId } = render(
      <Container>
        <NewsScreen />
      </Container>,
    )
    const req = MockAxios.getReqByMatchUrl(/navigation-by-hierarchy/)
    const category = { display_name: 'Política', url: '/politics/', _id: 'id' }
    const spy = jest.spyOn(Alert, 'alert')
    MockAxios.mockResponse(MOCK_RESPONSE)
    MockAxios.mockResponse({ data: { children: [category] } }, req)
    await act(async () => undefined)

    const icon = getByTestId('icon-home')
    const { name } = icon.props
    fireEvent.press(icon)
    const [, , buttons] = spy.mock.calls[0]
    buttons?.[1].onPress?.()
    await act(async () => undefined)
    expect(getByTestId('icon-home').props.name).not.toBe(name)
  })

  it('should show notification when section is added or removed', async () => {
    App.key = 'gestion'

    const fnShow = jest.fn()
    mockUseNotification.mockImplementation(() => ({
      show: fnShow,
    }))

    const { getByTestId } = render(
      <Container>
        <NewsScreen />
      </Container>,
    )
    const req = MockAxios.getReqByMatchUrl(/navigation-by-hierarchy/)
    const category = { display_name: 'Política', url: '/politics/', _id: 'id' }
    MockAxios.mockResponse(MOCK_RESPONSE)
    MockAxios.mockResponse({ data: { children: [category] } }, req)
    await act(async () => undefined)

    const icon = getByTestId('icon-home')
    fireEvent.press(icon)
    await act(async () => undefined)
    expect(fnShow).toBeCalledWith(expect.objectContaining({ type: 'success' }))
  })
})
