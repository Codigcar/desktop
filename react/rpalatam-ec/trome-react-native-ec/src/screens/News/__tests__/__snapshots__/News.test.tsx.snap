// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`News screen page should render card and ad 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "display": "flex",
          "flex": 1,
        },
      ]
    }
    testID="News-screen"
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderBottomColor": "#FFCB05",
            "borderBottomWidth": 1,
            "display": "flex",
            "elevation": 3,
            "shadowColor": "#050505",
            "shadowOffset": Object {
              "height": 1,
              "width": 0,
            },
            "shadowOpacity": 0.16,
            "shadowRadius": 2.22,
            "zIndex": 1,
          },
          Array [
            Object {
              "paddingTop": 0,
            },
            undefined,
          ],
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "display": "flex",
              "flexDirection": "row",
              "height": 64,
              "paddingHorizontal": 8,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "left": 8,
                "position": "absolute",
                "zIndex": 1,
              },
            ]
          }
        >
          <RNGestureHandlerButton
            collapsable={false}
            hitSlop={
              Object {
                "bottom": 4,
                "left": 4,
                "right": 4,
                "top": 4,
              }
            }
            onGestureEvent={[Function]}
            onGestureHandlerEvent={[Function]}
            onGestureHandlerStateChange={[Function]}
            onHandlerStateChange={[Function]}
            rippleColor={0}
            testID="goback-button"
          >
            <View
              accessible={true}
              hitSlop={
                Object {
                  "bottom": 4,
                  "left": 4,
                  "right": 4,
                  "top": 4,
                }
              }
              style={Object {}}
            >
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "height": 16,
                      "width": 16,
                    },
                  ]
                }
              >
                <SvgMock
                  fill="#C1C1C1"
                  height={16}
                  width={16}
                />
              </View>
            </View>
          </RNGestureHandlerButton>
        </View>
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "flex": 1,
                "marginLeft": 32,
                "paddingLeft": 4,
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "#665202",
                  "display": "flex",
                  "fontFamily": "NotoSans",
                  "fontSize": 18,
                  "fontWeight": "900",
                  "lineHeight": 28,
                  "textAlign": "left",
                },
                Object {
                  "fontFamily": "NotoSans-Black",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            politics
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "position": "absolute",
                "right": 8,
                "zIndex": 1,
              },
            ]
          }
        />
      </View>
    </View>
    <RCTScrollView
      ItemSeparatorComponent={[Function]}
      ListEmptyComponent={
        <View
          flex={1}
          pt="1"
          px="0.75"
        >
          <CardPlaceholder
            repeat={1}
            variant="default"
          />
          <CardPlaceholder
            repeat={5}
            variant="magazine"
          />
        </View>
      }
      ListFooterComponent={[Function]}
      data={
        Array [
          Story {
            "headline": "title_one",
            "id": "1",
            "last_updated_date": "date",
            "promo_items": Object {},
            "reading_time": undefined,
            "restrictions": undefined,
            "section": undefined,
            "subheadline": "description_one",
            "type": "basic",
            "url": "url",
          },
          Advertising {
            "id": "/28253241/elcomercio/pwa/sect/default/caja1",
            "size": "mediumRectangle",
            "sizes": Array [
              "mediumRectangle",
              "largeBanner",
              "banner",
            ],
          },
          Story {
            "headline": "title_two",
            "id": "2",
            "last_updated_date": "date",
            "promo_items": Object {},
            "reading_time": undefined,
            "restrictions": undefined,
            "section": undefined,
            "subheadline": "description_two",
            "type": "basic",
            "url": "url",
          },
        ]
      }
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onEndReachedThreshold={0.5}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      refreshControl={
        <RefreshControlMock
          colors={
            Array [
              "#0089ff",
            ]
          }
          onRefresh={[Function]}
          progressBackgroundColor="#FFFFFF"
          refreshing={false}
        />
      }
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={0.5}
      scrollIndicatorInsets={
        Object {
          "right": 1,
        }
      }
      stickyHeaderIndices={Array []}
      viewabilityConfigCallbackPairs={Array []}
    >
      <RCTRefreshControl />
      <View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flex": 1,
                  "paddingVertical": 16,
                },
              ]
            }
          >
            <RNGestureHandlerButton
              collapsable={false}
              onGestureEvent={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              onHandlerStateChange={[Function]}
              rippleColor={0}
              testID="card-action-area"
            >
              <View
                accessible={true}
                collapsable={false}
                nativeID="animatedComponent"
                style={Object {}}
              >
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flex": 1,
                        "paddingHorizontal": 16,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#2e3443",
                          "display": "flex",
                          "fontFamily": "NotoSerif",
                          "fontSize": 18,
                          "fontWeight": "700",
                          "lineHeight": 26,
                        },
                        Object {
                          "fontFamily": "NotoSerif",
                        },
                      ]
                    }
                    testID="card-title"
                  >
                    title_one
                  </Text>
                  <View
                    style={
                      Array [
                        Object {
                          "display": "flex",
                          "paddingTop": 8,
                        },
                      ]
                    }
                  >
                    <Text
                      numberOfLines={3}
                      style={
                        Array [
                          Object {
                            "color": "#6B7280",
                            "display": "flex",
                            "fontFamily": "NotoSans",
                            "fontSize": 14,
                            "lineHeight": 20,
                          },
                          Object {
                            "fontFamily": "NotoSans-Regular",
                          },
                        ]
                      }
                    >
                      description_one
                    </Text>
                  </View>
                </View>
              </View>
            </RNGestureHandlerButton>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "paddingHorizontal": 16,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#6B7280",
                          "display": "flex",
                          "fontFamily": "NotoSans",
                          "fontSize": 14,
                          "lineHeight": 20,
                        },
                        Object {
                          "fontFamily": "NotoSans-Regular",
                        },
                      ]
                    }
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flexDirection": "row",
                      },
                    ]
                  }
                >
                  <RNGestureHandlerButton
                    collapsable={false}
                    onGestureEvent={[Function]}
                    onGestureHandlerEvent={[Function]}
                    onGestureHandlerStateChange={[Function]}
                    onHandlerStateChange={[Function]}
                    rippleColor={0}
                    testID="card-save-button"
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      nativeID="animatedComponent"
                      style={Object {}}
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "display": "flex",
                              "padding": 4,
                            },
                          ]
                        }
                      >
                        <Icon
                          color="#6B7280"
                          name="bookmark"
                          size={22}
                        />
                      </View>
                    </View>
                  </RNGestureHandlerButton>
                  <RNGestureHandlerButton
                    collapsable={false}
                    onGestureEvent={[Function]}
                    onGestureHandlerEvent={[Function]}
                    onGestureHandlerStateChange={[Function]}
                    onHandlerStateChange={[Function]}
                    rippleColor={0}
                    testID="card-share-button"
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      nativeID="animatedComponent"
                      style={Object {}}
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "display": "flex",
                              "padding": 4,
                            },
                          ]
                        }
                      >
                        <Icon
                          color="#6B7280"
                          name="share-variant"
                          size={22}
                        />
                      </View>
                    </View>
                  </RNGestureHandlerButton>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#E5E5E5",
                  "display": "flex",
                  "height": 1,
                  "marginHorizontal": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "display": "flex",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  undefined,
                  undefined,
                ]
              }
            >
              <Banner
                adSize="mediumRectangle"
                adUnitID="/6499/example/banner"
                onAdLoaded={[Function]}
                validAdSizes={
                  Array [
                    "mediumRectangle",
                    "largeBanner",
                    "banner",
                  ]
                }
              />
            </View>
          </View>
          <View
            style={
              Array [
                Object {
                  "backgroundColor": "#E5E5E5",
                  "display": "flex",
                  "height": 1,
                  "marginHorizontal": 16,
                },
              ]
            }
          />
        </View>
        <View
          onLayout={[Function]}
          style={null}
        >
          <View
            style={
              Array [
                Object {
                  "display": "flex",
                  "flex": 1,
                  "paddingVertical": 16,
                },
              ]
            }
          >
            <RNGestureHandlerButton
              collapsable={false}
              onGestureEvent={[Function]}
              onGestureHandlerEvent={[Function]}
              onGestureHandlerStateChange={[Function]}
              onHandlerStateChange={[Function]}
              rippleColor={0}
              testID="card-action-area"
            >
              <View
                accessible={true}
                collapsable={false}
                nativeID="animatedComponent"
                style={Object {}}
              >
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flex": 1,
                        "paddingHorizontal": 16,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      Array [
                        Object {
                          "display": "flex",
                          "flexDirection": "row",
                        },
                      ]
                    }
                  >
                    <View
                      style={
                        Array [
                          Object {
                            "display": "flex",
                            "flex": 1,
                          },
                        ]
                      }
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#2e3443",
                              "display": "flex",
                              "fontFamily": "NotoSerif",
                              "fontSize": 18,
                              "fontWeight": "700",
                              "lineHeight": 26,
                            },
                            Object {
                              "fontFamily": "NotoSerif",
                            },
                          ]
                        }
                        testID="card-title"
                      >
                        title_two
                      </Text>
                    </View>
                  </View>
                </View>
              </View>
            </RNGestureHandlerButton>
            <View
              style={
                Array [
                  Object {
                    "display": "flex",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "display": "flex",
                      "flexDirection": "row",
                      "justifyContent": "space-between",
                      "paddingHorizontal": 16,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "display": "flex",
                        "flexDirection": "row",
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#6B7280",
                          "display": "flex",
                          "fontFamily": "NotoSans",
                          "fontSize": 14,
                          "lineHeight": 20,
                        },
                        Object {
                          "fontFamily": "NotoSans-Regular",
                        },
                      ]
                    }
                  />
                </View>
                <View
                  style={
                    Array [
                      Object {
                        "display": "flex",
                        "flexDirection": "row",
                      },
                    ]
                  }
                >
                  <RNGestureHandlerButton
                    collapsable={false}
                    onGestureEvent={[Function]}
                    onGestureHandlerEvent={[Function]}
                    onGestureHandlerStateChange={[Function]}
                    onHandlerStateChange={[Function]}
                    rippleColor={0}
                    testID="card-save-button"
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      nativeID="animatedComponent"
                      style={Object {}}
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "display": "flex",
                              "padding": 4,
                            },
                          ]
                        }
                      >
                        <Icon
                          color="#6B7280"
                          name="bookmark"
                          size={22}
                        />
                      </View>
                    </View>
                  </RNGestureHandlerButton>
                  <RNGestureHandlerButton
                    collapsable={false}
                    onGestureEvent={[Function]}
                    onGestureHandlerEvent={[Function]}
                    onGestureHandlerStateChange={[Function]}
                    onHandlerStateChange={[Function]}
                    rippleColor={0}
                    testID="card-share-button"
                  >
                    <View
                      accessible={true}
                      collapsable={false}
                      nativeID="animatedComponent"
                      style={Object {}}
                    >
                      <View
                        style={
                          Array [
                            Object {
                              "display": "flex",
                              "padding": 4,
                            },
                          ]
                        }
                      >
                        <Icon
                          color="#6B7280"
                          name="share-variant"
                          size={22}
                        />
                      </View>
                    </View>
                  </RNGestureHandlerButton>
                </View>
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "height": 0,
            }
          }
        />
        <View
          onLayout={[Function]}
        />
      </View>
    </RCTScrollView>
  </View>
</RNCSafeAreaProvider>
`;
